File: 234.回文链表.go
/*
 * @lc app=leetcode.cn id=234 lang=golang
 *
 * [234] 回文链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {
	mid := findMiddleNode(head)
	p := head
	q := mid.Next
	mid.Next = nil

	q = reverseNodeList(q)

	for p != nil && q != nil {
		if p.Val != q.Val {
			return false
		}

		p = p.Next
		q = q.Next
	}

	return true
}

func reverseNodeList(head *ListNode) *ListNode {
	var prev, current *ListNode
	current = head

	for current != nil {
		next := current.Next
		current.Next = prev
		prev = current
		current = next
	}

	return prev
}

func findMiddleNode(head *ListNode) *ListNode {
	slow, fast := head, head.Next
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	return slow
}

// @lc code=end

File: 234.回文链表.py
#
# @lc app=leetcode.cn id=234 lang=python3
#
# [234] 回文链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        mid = self.findMiddleNode(head)
        p = mid.next
        mid.next = None
        p = self.reverseNodeList(p)

        while head and p:
            if head.val != p.val:
                return False
            head = head.next
            p = p.next
        return True

    def findMiddleNode(self, head: Optional[ListNode]) -> ListNode:
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow

    def reverseNodeList(self, head: Optional[ListNode]) -> ListNode:
        prev, current = None, head
        while current:
            next = current.next
            current.next = prev
            prev = current
            current = next
        return prev

# @lc code=end

